### ISSUE TITLE
Add rule to detect reads of write-only-properties

### ISSUE THREAD

**Issue by ajafff** [fimbullinter/wotan#372]
TypeScript doesn't have a concept to support write-only properties in the same way it supports readonly properties. It treats properties with a setter as read/writable.

This can lead to bugs if you try to read the value of the property as it will always yield `undefined` but TypeScript infers another type.

```ts
const obj = new class {
  set prop(v: any) {}
}
console.log(obj.prop); // error here
obj.prop++; // error here, calls setter with NaN
```

TypeScript issue: https://github.com/Microsoft/TypeScript/issues/11596

TBD: how to handle unions of different class instances where the accessed property is write-only in one instance.

**Comment by ajafff** [comment]
This should also detect spreading objects with write-only properties:

```js
({
    foo: 1,
    ...{set foo(v) {}}
}).foo === undefined;
```

https://github.com/Microsoft/TypeScript/pull/26287#pullrequestreview-144829626

**Comment by ajafff** [comment]
This might be unnecessary if https://github.com/Microsoft/TypeScript/pull/26338 lands upstream

**Comment by DavidKDeutsch** [comment]
> This might be unnecessary if [microsoft/TypeScript#26338](https://github.com/microsoft/TypeScript/pull/26338) lands upstream

So microsoft/TypeScript#26338 has been closed with this comment by @RyanCavanaugh

> Closing because this can't be merged unless #21759 happens (which is quite unlikely in the near future)

So could we resurrect this feature request (i.e. detecting reads of write-only properties)? I just got bit by this, as I forgot to add a getter for one of my properties. I was actually pretty shocked that I did not get a compile error for the code that attempted to read it, as it seems to be a pretty straightforward check.

**Comment by ajafff** [comment]
This feature request became even more useful now that declaration files contain accessors instead of plain properties. That means it now works across package boundaries.

Will try to get this implemented in the near future.

**Comment by ajafff** [comment]
> Will try to get this implemented in the near future.

Seems like the "near future" is now, and it's not even the new year yet. PR is up at #674 

![image](https://user-images.githubusercontent.com/11968040/71635311-b1ff5680-2c23-11ea-8715-c9a7d0ea58da.png)

**Comment by ajafff** [comment]
:tada: This issue has been resolved in version [v0.23.0](https://github.com/fimbullinter/wotan/releases/tag/v0.23.0) :tada:
